### PR TITLE
Add support for crystal language.

### PR DESCRIPTION
**Pull request by hugopl**
The parser still under (active) development, but it's pretty usable, fast and doesn't crashed since I started using it.

- I use it daily for work and pet projects
- Indent not added yet because I didn't find the documentation about the captures, but I should add it soon.
- Locals not added because I didn't test it yet.

image

### CODE REVIEW

**Comment by hugopl** [comment]
Seems I have to fix some linter errors and add a test.

**Comment by hugopl** [comment]
No idea why it failed on Windows, could someone give me some directions?

**Comment by clason** [comment]
Before you spend more time on this, be aware that we no longer accept "experimental" or work in progress parsers. It needs to be stable both in terms of not crashing and in terms of not expecting a lot of breaking changes; we also expect some active maintainership of the queries here. 

(I believe a previous attempt at adding this language failed on that account; please check closed PRs.)

**Comment by hugopl** [comment]
Hi, last attempt failed because I did the PR targeting main instead of master.

I plan to maintain the queries since I use this daily at work and pet projects.

The parser isn't complete yet, so I'll change the PR to draft again and wait until it gets more mature. I'm also a contributor to the parser project, so will be easy to know when I can consider it ready to be merged here, meanwhile the parser README can be updated with instructions on how to manually add it to nvim.

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -325,6 +325,15 @@ list.cpp = {
   maintainers = { "@theHamsta" },
 }
 
+list.crystal = {
+  install_info = {
+    url = "https://github.com/crystal-lang-tools/tree-sitter-crystal",
+    files = { "src/parser.c", "src/scanner.c", "src/unicode.c" },
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:331]:
```suggestion
    files = { "src/parser.c", "src/scanner.c", "src/unicode.c" },
```

That is not really portable. Can you test that `tree-sitter build -o crystal.so` works correctly?

**Comment by clason** [comment]
How long do you expect the experimental phase to last? Having a draft PR up for months is not great.

**Comment by clason** [comment]
Sure, if you want, but there's automation that takes care of it. If you do, make sure all(!) lists remain alphabetized.

File: lua/nvim-treesitter/parsers.lua (hunk 1 above)
Comment (keidax) [reply lua/nvim-treesitter/parsers.lua:331]: @hugopl `unicode.c` doesn't need to be in this list, `scanner.c` actually just includes it inline


@clason 
The "Compile parsers" step on Windows includes these logs:
```
Downloading tree-sitter-crystal...
Creating temporary directory
Extracting tree-sitter-crystal...
Compiling...
Treesitter parser for crystal has been installed
```
Do you think that's failing silently?

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:331]: Yes, since it then complains about the parser not being installed.

**Comment by hugopl** [comment]
> How long do you expect the experimental phase to last? Having a draft PR up for months is not great.

I'm using this every day, so I think even the parser not being complete yet, what we have now is good enough and the crystal community really lack editor support, that said, I plan to update the PR next days and mark it as ready for review.

File: lua/nvim-treesitter/parsers.lua (hunk 1 above)
Comment (hugopl) [reply lua/nvim-treesitter/parsers.lua]: Fixed.

**Comment by clason** [comment]
I'm mostly worried about how frequent (breaking) changes will be.

**Comment by hugopl** [comment]
> I'm mostly worried about how frequent (breaking) changes will be.

No need to worry about that, since I'm also part of tree-sitter-crystal, follow its development closely and use this everyday on my daily job I have an huge interest in see this always working, plus the lock file help to mitigate these breakages.

However I have no idea about why the windows CI is failing and would need some help with that.

**Comment by clason** [comment]
>  the lock file help to mitigate these breakages.

No, that's not quite so simple. A breaking change to the parser will block ALL updates and requires manual intervention. I don't want to deal with that daily or weekly.

> However I have no idea about why the windows CI is failing and would need some help with that.

I would guess it's the `unicode.c`. As I don't have access to a Windows machine (and no experience with building there), I can't help you unfortunately.

**Comment by keidax** [comment]
@hugopl I thought https://github.com/crystal-lang-tools/tree-sitter-crystal/pull/61would fix the Windows failure, maybe there's another error we're not seeing here :thinking: 

I'm happy to dig into that further, I'll have time over the next few days.

**Comment by keidax** [comment]
> I'm mostly worried about how frequent (breaking) changes will be.

As the main author of tree-sitter-crystal, I don't think the grammar is stable enough to merge here quite yet.

The grammar is mostly complete, except for handling Crystal macros. I'm planning to work on macro support very soon, but it may require breaking changes.

My hope is that we're a few weeks away from an initial stable release of tree-sitter-crystal.

> >the lock file help to mitigate these breakages.

> No, that's not quite so simple. A breaking change to the parser will block ALL updates and requires manual intervention. I don't want to deal with that daily or weekly.

Can you explain what this means? If an upstream parser releases a breaking change, but the lockfile in the project stays the same, what is the impact?

**Comment by clason** [comment]
> Can you explain what this means? If an upstream parser releases a breaking change, but the lockfile in this project stays the same, what is the impact?

The lockfile is updated by an automated action, which will automatically merge non-breaking updates but fail if there are breaking ones, requiring manual intervention. As more and more parsers are added, this is scaling increasingly poorly, so we need to be increasingly conservative.

**Comment by hugopl** [comment]
> As the main author of tree-sitter-crystal, I don't think the grammar is stable enough to merge here quite yet.

That said, I'll close this one and add instructions on how to use tree-sitter-crystal on nvim on tree-sitter-crystal README.

Later I can open another PR.

**Pull request closed without merging**